spring-core/src/main/java/org/springframework/util/UnmodifiableMultiValueMap.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.util;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.lang.Nullable;
+
+
+/**
+ * Immutable implementation of {@link MultiValueMap}.
+ *
+ * @author Brian Clozel
+ * @since 5.1
+ * @param <K> the key type
+ * @param <V> the value element type
+ */
+public class UnmodifiableMultiValueMap<K, V> implements MultiValueMap<K, V> {
```

Comment:
I don't think this needs to be a public type. It could simply be package friendly, and returned from org.springframework.util.CollectionUtils#unmodifiableMultiValueMap.

But I'm not even sure we need it at all, as I'd rather see a `ReadOnlyHttpHeaders`, see below.